Avoid some max constructor optimizations when known not needed.

laithsakka opened the pull request:
Summary: 
around 10% with 1K nodes
more than that with 2K features. 414.5735 -> 333 (20%)

This target optimizing patterns like this
```
 sym_max: "Sym(Max(u31 + u32, u33 + u34))" = torch.sym_max(sym_sum_6, sym_sum_7);  sym_sum_6 = sym_sum_7 = None
        sym_max_1: "Sym(Max(u31 + u32, u33 + u34, u35 + u36))" = torch.sym_max(sym_max, sym_sum_8);  sym_max = sym_sum_8 = None
        sym_max_2: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38))" = torch.sym_max(sym_max_1, sym_sum_9);  sym_max_1 = sym_sum_9 = None
        sym_max_3: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40))" = torch.sym_max(sym_max_2, sym_sum_10);  sym_max_2 = sym_sum_10 = None
        sym_max_4: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40, u41 + u42))" = torch.sym_max(sym_max_3, sym_sum_11);  sym_max_3 = sym_sum_11 = None
        sym_max_5: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40, u41 + u42, u43 + u44))" = torch.sym_max(sym_max_4, sym_sum_12);  sym_max_4 = sym_sum_12 = None
        sym_max_6: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40, u41 + u42, u43 + u44, u45 + u46))" = torch.sym_max(sym_max_5, sym_sum_13);  sym_max_5 = sym_sum_13 = None
        sym_max_7: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40, u41 + u42, u43 + u44, u45 + u46, u47 + u48))" = torch.sym_max(sym_max_6, sym_sum_14);  sym_max_6 = sym_sum_14 = None
        sym_max_8: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40, u41 + u42, u43 + u44, u45 + u46, u47 + u48, u49 + u50))" = torch.sym_max(sym_max_7, sym_sum_15);  sym_max_7 = sym_sum_15 = sym_max_8 = None
```


<img width="496" alt="Screenshot 2024-11-05 at 11 00 35 AM" src="https://github.com/user-attachments/assets/455c06a3-e1bf-43cb-b880-9470ae6fb07f">
<img width="511" alt="Screenshot 2024-11-05 at 11 00 57 AM" src="https://github.com/user-attachments/assets/ff0d4236-9b5c-4a9a-8520-47b005bb3cb0">

Differential Revision: D65354971




cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

ezyang commented on torch/_dynamo/config.py:
```diff
@@ -520,6 +520,8 @@ def default_debug_dir_root():
     "TORCH_DYNAMO_AUTOMATIC_DYNAMIC_REMOTE_PGO"
 )
 
+extra_validations = False
```
Needs more docs. Unclear if you've named it a long enough / unique enough name.

laithsakka replied:
I will add a comment, I am thinking

# This is turned on all dynamo tested, can be used to wrap expensive assertion and checks to run on all dynamo unit tests.

bobrenjc93 replied:
@laithsakka I'd recommend using something like chatgpt/metamate to correct your grammar, especially for these documentation related comments. Personally I'd err on the side of specificity, something like


```
# This config option enables additional sympy validations to verify correctness, though it comes at a performance cost.
# For example, in some cases, we skip expensive function calls, such as `_collapse_arguments` and `_find_localzeros`.
# When this option is enabled, we perform these additional function calls and verify that they are indeed unnecessary 
# by asserting that expressions with and without these passes are equivalent. Currently we only flip this flag
# in our dynamo tests.
do_extra_sympy_validations = False
```

laithsakka replied:
I will run my comment on chat GPT, I intend this config go be general and not just tied to my validations. 
basically any validation that someone want to add and wanted to run during dynamo test.

laithsakka replied:
I will update it to elaborate more.

laithsakka replied:
added more details and run it by metamate

isuruf commented:
Do you have a code that I can run that this PR helps with?

laithsakka commented:
> [isuruf](/isuruf)

i also have this https://github.com/pytorch/pytorch/pull/139867 and some other opts following

laithsakka commented:
> Do you have a code that I can run that this PR helps with?

i will actually also in addition to the above add a pr time benchmark that reflect those modele behaviors

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -562,23 +562,37 @@ class MinMaxBase(Expr, LatticeOp):  # type: ignore[misc]
 
-        evaluate = assumptions.pop("evaluate", global_parameters.evaluate)
-        args = (sympify(arg) for arg in args)
+        import torch._dynamo.config as config
```
There's some hokey layering violation but... fine.

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -562,23 +562,37 @@ class MinMaxBase(Expr, LatticeOp):  # type: ignore[misc]
-        # first standard filter, for cls.zero and cls.identity
-        # also reshape Max(a, Max(b, c)) to Max(a, b, c)
+        evaluate = assumptions.pop("evaluate", global_parameters.evaluate)
+        args = [sympify(arg) for arg in args]
```
became a list instead of generator 🤔

laithsakka replied:
was trying to avid typing error but this even did not fix it, but since it did not introduce regression i kept it as is, but i will revert it back . this is not needed.

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -589,8 +603,23 @@ def __new__(cls, *args, **assumptions):
+        """Return true if all atoms in all args are all unique and there are no constants atoms"""
+        seen_symbols = set()
+        for arg in args:
+            for element in arg.atoms():
```
The use of atoms here seems suspicious. Why is it ok to iterate over atoms?

laithsakka replied:
i can restrict this to summations of symbols that are unique, but my intuition was the following:

iterating over atoms above implies the following:
1. There is no constants in any input expression.
2. There are no shared symbols across expressions. 

that applies (no proof here)
1. We cant tell that some arg is bigger or smaller than another (well because they are made of complete different symbols and there is no numeric values on each side). 
2. We can have redundant expressions or common elements (same reasoning above) 

so I could not think of max(e1, e2)
where e1 and e2 do not have numbers, do not share symbols but i can tell that one is greater than the other. 

Finally those opts sounded like best effort anyway from the comment above (remove redundant args that are easily identified). so i thought in the worse case if i missed one, its clearly a complicated one !

let me know if that convince you , that said I am happy to restrict this to summations of symbols of form (Max(a+b, c+d, ..)).

laithsakka replied:
One more thing is its also assumed inputs are already simplified , none of them is (a-a) or (b-b) which is valid assumtion.

ezyang replied:
It is difficult to see without an example.

It seems I can generate a synthetic function that will cause you to go wrong. Let MulByZero be a function which multiplies by zero. We now have Min(MulByZero(s0), MulByZero(s1)). The atoms of MulByZero(s0) are simply s0. By the logic here, there are no non-symbol atoms and therefore everything is unique. But clearly I can simplify this whole shebang to zero.

In general, when you reason about atoms, you are saying that whenever I have Min(F1(s0, ...), F2(s1, ...)), *NO MATTER* the choice of F1/F2, there is guaranteed not to be any possible simplification from collapse arguments / localzeros. This seems difficult to prove!

laithsakka replied:
I mean MulByZero is synthetic function, has it been x*0, it would have been 0, since inputs are already simplified .

I understand that this is difficult to prove and in fact not correct for all type of expressions that anyone can create, but it sounds to me that it would work with the type of expressions that we deal with in pytorch here. run_extra_validations will catch if we ran anything in Pytorch that would invalidate that (in the tests).

That said, since in this benchmark we are dealing with summations inputs. I will just rewrite it to make stricter to that.

ezyang replied:
So, just to recap discussion from offsite, we were *only* going to look into summations (and potentially polynomials too), because that's what you actually need in practice and otherwise it's hard to reason through what exactly the optimization here is.

Once you limit to summations, you can just say: "Let's suppose we are doing Max(polynomial1, polynomial2)". When the symbols of polynomial1 and polynomial2 are distinct, we know it's not possible for there to be any potential simplification to the formula here, so we bypass simplification."

By the way, a *better* (i.e., more permanent) approach is to take the overly general preexisting simplification logic in Sympy and figure out what part of it actually matters for ML workloads, and then rewrite all of it to just do that simple core (which is also fast!) This is a bit harder to do though because you have to figure out what actually matters and what doesn't, so there may be some trial and error.

ezyang commented on benchmarks/dynamo/pr_time_benchmarks/benchmarks/symint_sum.py:
```diff
@@ -27,17 +34,25 @@ def _work(self):
+    ).enable_compile_time_instruction_count().collect_all().append_results(result_path)
+    Benchmark(
+        use_loop=True
+    ).enable_compile_time_instruction_count().collect_all().append_results(result_path)
```
So now it's impossible to run only one of these benchmarks? lol

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -89,6 +89,17 @@
+        and len(expr._args) == 2
+        and expr._args[0].is_symbol
+        and expr._args[1].is_symbol
+        and expr._args[0] != expr._args[1]
```
nit: `expr._args[0] is not expr._args[1]`

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -559,26 +570,43 @@ def eval(cls, base, shift):
+                args_test = frozenset(args)
+                assert Expr.__new__(
+                    cls, *ordered(args_test), **assumptions
+                ) == Expr.__new__(cls, *ordered(args), **assumptions)
```
nit: print the before after expr when this assert fails

ezyang replied:
More broadly, to be honest, I'm not really sure how useful this assert actually will be. At the slightest whiff of trouble, where this assert fails but the results are still semantically correct, I will probably just dump this assert.

laithsakka replied:
i will just remove it

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -559,26 +570,43 @@ def eval(cls, base, shift):
-            args = cls._find_localzeros(args, **assumptions)
 
-        args = frozenset(args)
+            # When this is not none we can bypass _find_localzeros and _collapse_arguments.
```
I think for future understandability, the comment here would be good to explain what "unique summation atoms" denotes. In particular, what does it mean when unique summation atoms is None? Why does this condition mean we should skip collapse arguments / find localzeros? Some of these answers can be found in the docstrings of the function call itself, but *this* is the logical place to explain how it all comes together.

laithsakka replied:
done

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -589,8 +617,53 @@ def __new__(cls, *args, **assumptions):
         return obj
 
+    @classmethod
+    def _should_optimize_max_of_unique_summations(cls, args):
```
Please annotate with input/return type (really, the return type is the important one here)

laithsakka replied:
done

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -589,8 +617,53 @@ def __new__(cls, *args, **assumptions):
+        """Optimization for patterns of the form
+          max(a+b, c+d)
+          max(max(a+b, c+d), e+f) = max(a+b, c+d, e+f) = k
+          max(k, g+h) .. =  max(a+b, c+d, e+f, g+h)..etc
```
It's not really clear what these equations are trying to show. You can explain it with more words which would make it clearer. Something like: "Sometimes, we successfully build larger and larger Max expressions where the arguments ... ... ..."

laithsakka replied:
addressed

ezyang commented on torch/utils/_sympy/functions.py:
```diff
@@ -589,8 +617,53 @@ def __new__(cls, *args, **assumptions):
         obj._argset = args
+
+        if unique_summation_atoms:
+            obj.unique_summation_atoms = unique_summation_atoms
```
Need documentation

ezyang replied:
The naming here is a not great. "unique summation atoms" suggests that this is some sort of precomputed value representing unique atoms that are being summed together. But what exactly does "unique" mean here (it means that if a symbol shows up multiple times, this is actually None); and what exactly are valid "atoms" (in fact, only symbols are valid, if you have constants this is None).

I'm not exactly sure what to name it. "Certificate" might be a good term to use, as the presence of this property implies that some property holds of the expression. You might want to just invent some term of art here, I can't really think of a good name of the property though.

Why not set this property unconditionally, instead of having to test hasattr on it? That seems like a simpler runtime invariant.

laithsakka replied:
I am unsure what naming i should use but i added comments to make things more clear  with example and did some refactor based on your comments. the naming is still not gr8 unique_summations_symbols" but cant think of a good name, however a big comment that i added explains the optimization in details

ezyang reviewed (approved):
I am approving this because (1) I think the change here is functionally correct and (2) I am reasonably confident I understand this code enough to be able to make changes to it in the future. But I think it is not so aligned with my vision of what "good" code in the symbolic reasoning system looks like, so I'm hoping my comments can be a way for us to get more aligned on this matter, esp for future changes.

laithsakka commented:
> I am approving this because (1) I think the change here is functionally correct and (2) I am reasonably confident I understand this code enough to be able to make changes to it in the future. But I think it is not so aligned with my vision of what "good" code in the symbolic reasoning system looks like, so I'm hoping my comments can be a way for us to get more aligned on this matter, esp for future changes.

i will address your comment before landing

laithsakka committed:
Avoid some max constructor optimizations when known not needed. (#139741)

Summary:

this combines both https://www.internalfb.com/diff/D65368242 and the original change. 
total savings is :349s -> 256s


this targets optimizing patterns like:
```"
 sym_max: "Sym(Max(u31 + u32, u33 + u34))" = torch.sym_max(sym_sum_6, sym_sum_7);  sym_sum_6 = sym_sum_7 = None
        sym_max_1: "Sym(Max(u31 + u32, u33 + u34, u35 + u36))" = torch.sym_max(sym_max, sym_sum_8);  sym_max = sym_sum_8 = None
        sym_max_2: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38))" = torch.sym_max(sym_max_1, sym_sum_9);  sym_max_1 = sym_sum_9 = None
        sym_max_3: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40))" = torch.sym_max(sym_max_2, sym_sum_10);  sym_max_2 = sym_sum_10 = None
        sym_max_4: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40, u41 + u42))" = torch.sym_max(sym_max_3, sym_sum_11);  sym_max_3 = sym_sum_11 = None
        sym_max_5: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40, u41 + u42, u43 + u44))" = torch.sym_max(sym_max_4, sym_sum_12);  sym_max_4 = sym_sum_12 = None
        sym_max_6: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40, u41 + u42, u43 + u44, u45 + u46))" = torch.sym_max(sym_max_5, sym_sum_13);  sym_max_5 = sym_sum_13 = None
        sym_max_7: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40, u41 + u42, u43 + u44, u45 + u46, u47 + u48))" = torch.sym_max(sym_max_6, sym_sum_14);  sym_max_6 = sym_sum_14 = None
        sym_max_8: "Sym(Max(u31 + u32, u33 + u34, u35 + u36, u37 + u38, u39 + u40, u41 + u42, u43 + u44, u4
```
 {F1953835308}{F1953835311}

Test Plan: run existing tests.

Reviewed By: ezyang, aorenste

Differential Revision: D65354971

laithsakka commented:
address all comments

laithsakka commented:
@pytorchbot merge
